src/main/scala/io/iohk/ethereum/merklepatriciatrie/MerklePatriciaTrie.scala

```diff
@@ -0,0 +1,486 @@
+
+  private def getNode(nodeId: Array[Byte], source: DataSource)(implicit nodeDec: RLPDecoder[Node]): Node = {
+    val nodeEncoded =
+      if (nodeId.length < 32) nodeId
+      else source.get(nodeId).getOrElse(throw MPTException("Node not found, trie is inconsistent"))
+    decodeRLP[Node](nodeEncoded)
+  }
+
+  private def matchingLength(a: Array[Byte], b: Array[Byte]): Int = a.zip(b).takeWhile(t => t._1 == t._2).length
+
+  private def updateNodesInStorage(previousRootHash: Array[Byte], newRootHash: Array[Byte], newRoot: Option[Node],
+                                   toRemove: Seq[Node], toUpdate: Seq[Node], dataSource: DataSource, hashFn: HashFn): DataSource = {
+    val rootCapped = newRoot.map(_.capped).getOrElse(Array.emptyByteArray)
+    val toBeRemoved = toRemove.filter { node =>
+      val nCapped = node.capped
+      nCapped.length == 32 || (node.hash sameElements previousRootHash)
+    }.map(_.hash)
+    val toBeUpdated = toUpdate.filter { n =>
+      val nCapped = n.capped
+      nCapped.length == 32 || (nCapped sameElements rootCapped)
+    }.map(node => node.hash -> node.encode)
+    dataSource.update(newRootHash, toBeRemoved, toBeUpdated)
+  }
+
+  private def getNextNode(extensionNode: ExtensionNode, dataSource: DataSource)(implicit nodeDec: RLPDecoder[Node]): Node =
+    extensionNode.next match {
+      case Right(node) => node
+      case Left(hash) => MerklePatriciaTrie.getNode(hash, dataSource)
+    }
+
+  private def getChild(branchNode: BranchNode, pos: Int, dataSource: DataSource)(implicit nodeDec: RLPDecoder[Node]): Option[Node] =
+    branchNode.children(pos) map {
+      case Right(node) => node
+      case Left(hash) => MerklePatriciaTrie.getNode(hash, dataSource)
+    }
+
+  private[merklepatriciatrie] implicit val nodeEnc = new RLPEncoder[Node] {
+    override def encode(obj: Node): RLPEncodeable = obj match {
+      case leaf: LeafNode => RLPList(HexPrefix.encode(nibbles = leaf.key, isLeaf = true), leaf.value)
+      case extension: ExtensionNode =>
+        RLPList(HexPrefix.encode(nibbles = extension.sharedKey, isLeaf = false), extension.next match {
+          case Right(node) => this.encode(node)
+          case Left(bytes) => bytes
+        })
+      case branch: BranchNode =>
+        val toEncode: Seq[RLPEncodeable] = branch.children.map {
+          case Some(Right(node)) => this.encode(node)
+          case Some(Left(bytes)) => RLPValue(bytes)
+          case None => RLPValue(Array.emptyByteArray)
+        } :+ RLPValue(branch.terminator.getOrElse(Array.emptyByteArray))
+        RLPList(toEncode: _*)
+    }
+  }
+
+  implicit val defaultByteArraySerializable = new ByteArraySerializable[Array[Byte]] {
+    override def toBytes(input: Array[Byte]): Array[Byte] = input
+
+    override def fromBytes(bytes: Array[Byte]): Array[Byte] = bytes
+  }
+}
+
+class MerklePatriciaTrie[K, V](private val rootHash: Option[Array[Byte]],
+                               val dataSource: DataSource,
```

Comment:
I think of MPT as any other Trie that instead of using an in memory structure to store it's info might (or not) use a persistent storage so ok, we agreed not to create Storage, MPT and DataSource.

About naming. In EthereumJ, go-ethereum and python versions it's not named under DB. So i'm not sure. I don't mind renaming it though.

ps: lol, I don' t know either what' Mut stands for.